File: create-meta/javascript/src/createMeta.ts
import os from 'os'
import { parse as parseUrl, format as formatUrl } from 'url'
import * as messages from '@cucumber/messages'
import defaultCiDict from './ciDict.json'
import evaluateVariableExpression from "./evaluateVariableExpression";

export type CiDict = Record<string,CiSystem>
export type Env = Record<string,string|undefined>

export interface CiSystem {
  url: string
  git: {
    remote: string | undefined
    branch: string | undefined
    revision: string | undefined
    tag: string | undefined
  }
}

export default function createMeta(
  toolName: string,
  toolVersion: string,
  envDict: Env,
  ciDict?: CiDict
): messages.Meta {
  if (ciDict === undefined) {
    ciDict = defaultCiDict
  }
  return {
    protocolVersion: messages.version,
    implementation: {
      name: toolName,
      version: toolVersion,
    },
    cpu: {
      name: os.arch(),
    },
    os: {
      name: os.platform(),
      version: os.release(),
    },
    runtime: {
      name: 'node.js',
      version: process.versions.node,
    },
    ci: detectCI(ciDict, envDict),
  }
}

export function detectCI(ciDict: CiDict, envDict: Env): messages.Ci | undefined {
  const detected: messages.Ci[] = []
  for (const [ciName, ciSystem] of Object.entries(ciDict)) {
    const ci = createCi(ciName, ciSystem, envDict)
    if (ci) {
      detected.push(ci)
    }
  }
  if (detected.length !== 1) {
    return undefined
  }
  if (detected.length > 1) {
    console.error(`@cucumber/create-meta WARNING: Detected more than one CI: ${JSON.stringify(detected, null, 2)}`)
    console.error('Using the first one.')
  }
  return detected[0]
}

export function removeUserInfoFromUrl(value: string): string {
  if (!value) return value
  const url = parseUrl(value)
  if (url.auth === null) return value
  url.auth = null
  return formatUrl(url)
}

function createCi(ciName: string, ciSystem: CiSystem, envDict: Env): messages.Ci | undefined {
  const url = evaluateVariableExpression(ciSystem.url, envDict)
  if (url === undefined) {
    // The url is what consumers will use as the primary key for a build
    // If this cannot be determined, we return nothing.
    return undefined
  }

  let branch = evaluateVariableExpression(ciSystem.git.branch, envDict);
  return {
    url,
    name: ciName,
    git: {
      remote: removeUserInfoFromUrl(evaluateVariableExpression(ciSystem.git.remote, envDict)),
      revision: evaluateVariableExpression(ciSystem.git.revision, envDict),
      branch: branch,
      tag: evaluateVariableExpression(ciSystem.git.tag, envDict),
    },
  }
}

File: create-meta/ruby/lib/cucumber/create_meta.rb
require 'uri'
require 'sys/uname'
require 'json'
require 'cucumber/messages'
require 'cucumber/create_meta/variable_expression'

module Cucumber
  module CreateMeta
    extend Cucumber::CreateMeta::VariableExpression
    CI_DICT = JSON.parse(IO.read(File.join(File.dirname(__FILE__), "ciDict.json")))

    def create_meta(tool_name, tool_version, env = ENV)
      Cucumber::Messages::Meta.new(
        protocol_version: Cucumber::Messages::VERSION,
        implementation: Cucumber::Messages::Product.new(
            name: tool_name,
            version: tool_version
        ),
        runtime: Cucumber::Messages::Product.new(
            name: RUBY_ENGINE,
            version: RUBY_VERSION
        ),
        os: Cucumber::Messages::Product.new(
            name: RbConfig::CONFIG['target_os'],
            version: Sys::Uname.uname.version
        ),
        cpu: Cucumber::Messages::Product.new(
            name: RbConfig::CONFIG['target_cpu']
        ),
        ci: detect_ci(env)
      )
    end

    def detect_ci(env)
      detected = CI_DICT.map do |ci_name, ci_system|
        create_ci(ci_name, ci_system, env)
      end.compact

      detected.length == 1 ? detected[0] : nil
    end

    def create_ci(ci_name, ci_system, env)
      url = evaluate(ci_system['url'], env)
      return nil if url.nil?

      Cucumber::Messages::Ci.new(
        url: url,
        name: ci_name,
        build_number: evaluate(ci_system['buildNumber'], env),
        git: Cucumber::Messages::Git.new(
          remote: remove_userinfo_from_url(evaluate(ci_system['git']['remote'], env)),
          revision: evaluate(ci_system['git']['revision'], env),
          branch: evaluate(ci_system['git']['branch'], env),
          tag: evaluate(ci_system['git']['tag'], env),
        )
      )
    end

    def remove_userinfo_from_url(value)
      return nil if value.nil?
      begin
        uri = URI(value)
        uri.userinfo = ''
        uri.to_s
      rescue
        value
      end
    end

    module_function :create_meta, :detect_ci, :create_ci, :remove_userinfo_from_url
  end
end
